src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py

```diff
@@ -196,39 +209,63 @@ def update_progress(cur_loss, iteration):
     model.hybridize()
     trainer = _mx.gluon.Trainer(model.collect_params(), 'adam')
 
-    train_loss = 0.
-    for batch in loader:
-        data = _mx.gluon.utils.split_and_load(batch.data[0], 
-            ctx_list=ctx, batch_axis=0)[0]
-        label = _mx.nd.array(
-            _mx.gluon.utils.split_and_load(batch.label[0], 
-                ctx_list=ctx, batch_axis=0)[0]
-            )
+    train_accuracy = _mx.metric.Accuracy()
+    validation_accuracy = _mx.metric.Accuracy()
+    for train_batch in train_loader:
+        train_batch_data = _mx.gluon.utils.split_and_load(
+            train_batch.data[0], ctx_list=ctx, batch_axis=0)
+        train_batch_label = _mx.gluon.utils.split_and_load(
+                train_batch.label[0], ctx_list=ctx, batch_axis=0)
+        train_outputs = []
 
+        # Inside training scope
         with _autograd.record():
-            output = model(data)
-            loss = softmax_cross_entropy(output, label)
-        loss.backward()
-        # update parameters
-        trainer.step(1)
+            for x, y in zip(train_batch_data, train_batch_label):
+                z = model(x)
+                # Computes softmax cross entropy loss.
+                loss = softmax_cross_entropy(z, y)
+                # Backpropagate the error for one iteration.
+                loss.backward()
+                train_outputs.append(z)
+            # Compute training accuracy
+            train_accuracy.update(train_batch_label, train_outputs)
+
+        # Compute validation accuracy
+        for validation_batch in validation_loader:
```

Comment:
Doesn't this iterate through the validation loader for _every_ mini-batch (training step)?

Reply:
Doesn't the validation_loader need to be reset? How does it iterate through the data more than once?

Reply:
You're right. The validation loader in fact runs the data through only one epoch throughout training. Fixing this now.